lib/req_snowflake/req_snowflake.ex

```diff
@@ -0,0 +1,303 @@
+
+  It uses the Snowflake REST API to communicate with Snowflake, with an earlier version set for JSON.
+  There isn't an Elixir Arrow library (yet!), so it seems that setting an earlier Java version seems
+  to give us back JSON results. The REST API is used by the Python, Golang and other languages to
+  send requests to Snowflake, so it is stable and shouldn't just randomly break.
+
+  It is not the [Snowflake SQL API](https://docs.snowflake.com/en/developer-guide/sql-api/index.html), which is
+  limited in its implementation.
+
+  Right now the library doesn't support [MFA](https://docs.snowflake.com/en/user-guide/security-mfa.html), so you'll
+  need to either use [private key auth](https://docs.snowflake.com/en/user-guide/odbc-parameters.html#using-key-pair-authentication) or
+  connect using a username & password. A private key auth is highly recommended as you can rotate passwords easier.
+
+  This module handles the connection to Snowflake.
+
+  The module also supports async queries, which means that the query will be sent to Snowflake, then polled for updates.
+  This means that there won't (hopefully) be long connection times, so deploying should be easier as you won't run the risk
+  of killing running queries.
+
+  Query results are decoded into the `ReqSnowflake.Result` struct.
+
+  The struct implements the `Table.Reader` protocol and thus can be efficiently traversed by rows or columns.
+  """
+
+  alias Req.Request
+  alias ReqSnowflake.Result
+  alias ReqSnowflake.Snowflake
+
+  @allowed_options ~w(snowflake_query username password account_name region warehouse role database schema query_tag rows_per_resultset statement_timeout_in_seconds application_name bindings)a
+
+  @doc """
+  Attaches to Req request, used for querying Snowflake.
+
+  ## Request Options
+  * `:account_name` - Required. Account name. This is usually the name between the https:// and us-east-1 (or whatever region).
+                      If unsure, run `select current_account();` in Snowflake.
+  * `:region` - Required. Your snowflake region, the region is found between the account name and ".snowflakecomputing.com" on the portal.
+                If unsure, run `select current_region();` in Snowflake to show it. Example is `us-east-1`. Region names and their
+                IDs can be [found here](https://docs.snowflake.com/en/user-guide/intro-regions.html)
+  """
+  @spec attach(Request.t(), keyword()) :: Request.t()
+  def attach(%Request{} = request, options \\ []) do
+    request
+    |> Request.prepend_request_steps(snowflake_run: &run/1)
+    |> Request.register_options(@allowed_options)
+    |> Request.merge_options(options)
+  end
+
+  # Bindings happen for inserts
+  defp run(
+         %Request{
+           options:
+             %{
+               account_name: account_name,
+               region: region,
+               snowflake_query: query,
+               bindings: bindings
+             } = options
+         } = request
+       ) do
```

Comment:
Be super careful when you're doing pattern matching with a "catch-all" clause at the end. You may accidentally hit that catch-all!

I'd change this to something like:

```elixir
  defp run(%Request{options: %{snowflake_query: query}} = request) do
```

because otherwise if someone passes `snowflake_query` but NOT any other options, this step would be a no-op. It should only be a no-op when we don't set the query.